### ISSUE TITLE
"E102 expl3 control sequences in non-expl3 parts" occurs if an 'doc' expl block occurs in a `.dtx`

### ISSUE THREAD

**Issue by josephwright** [Witiko/expltools#8]
In [`siunitx-unit.dtx`](https://github.com/josephwright/siunitx/blob/main/siunitx-unit.dtx), I have one command set up in the preamble:
```latex
\ExplSyntaxOn
\makeatletter
\NewDocumentCommand \acro { m }
  {
    \textsc
      {
        \exp_args:NV \tl_if_head_eq_charcode:nNTF \f@series { m }
          { \text_lowercase:n }
          { \use:n }
            {#1}
      }
  }
\makeatother
\ExplSyntaxOff
```
This leads to  776 errors for the file: all E102. Removing that block fixes the error - so something is being missed.

**Comment by josephwright** [comment]
My guess is that after `\ExplSyntaxOff`, the `\ProvidesExplPackage` is not being picked up

**Comment by Witiko** [comment]
Many thanks for the early bug reports and sorry for the false alarms. I should use `siunitx-unit.dtx` as one of my regression tests, so that I can keep my eye on it as I add support for more warnings and errors.

**Comment by Witiko** [comment]
@josephwright: Removing the block from your original post actually turns the 776 errors into the following warning:

> siunitx-unit.dtx:                                     W100 no standard delimiters

This is because no other delimiters like `\ExplSyntaxOn` and `\ExplSyntaxOff` are present in the rest of the file `siunitx-unit.dtx` outside that block. Therefore, `explcheck` assumes that all code in the file is in expl3 but it issues a warning in case it got that assumption wrong.

> My guess is that after `\ExplSyntaxOff`, the `\ProvidesExplPackage` is not being picked up

It would have been picked up. For example, if you put the following line after `\ExplSyntaxOff`, the warning will disappear:

``` tex
\ProvidesExplFile{example.tex}{2024-04-09}{1.0.0}{An example file}
```

However, I can't see `\ProvidesExplPackage` anywhere in the file `siunitx-unit.dtx`. Could that be the issue?

**Comment by josephwright** [comment]
> However, I can't see `\ProvidesExplPackage` anywhere in the file `siunitx-unit.dtx`. Could that be the issue?

I guess :)

`siunitx` uses a sert of `.dtx` files that are DocStripped into one, so only `siunitx.dtx` has `\ProvidesExpl...`. But only the `unit` source cuaes an issue.

**Comment by Witiko** [comment]
> files [...] are DocStripped into one

That could be part of the issue. In general, explcheck can't make much sense of .dtx files that contain several DocStrip streams. In the file `siunitx.dtx`, it is confused by an `\ExplSyntaxOff` delimiter in the `driver` stream. However, that stream is used to typeset the package documentation and has little relevance to the rest of the code in the .dtx file. Running it on the final .sty file should do the trick for the moment.

> But only the unit source c[aus]es an issue.

There seems to be no expl3 code in the `driver` stream in your other .dtx files.

**Comment by Witiko** [comment]
> Running explcheck on the final .sty file should do the trick for the moment. Admittedly, this will make the line numbers in the warning and error messages more difficult to interpret.

Proper support for DocStrip-style delimiters in .dtx files is a useful feature but not planned for the next bugfix release. Therefore, I unassigned it from the corresponding milestone and marked this ticket as a feature request.

**Comment by Witiko** [comment]
While `.dtx` streams provide a structured way to analyze code, they may not align with the actual files that `explcheck` should process. An `.ins` file can combine multiple streams from different `.dtx` files to generate `.tex`, `.sty`, or `.cls` files, which are the real targets for static analysis. For instance, in your example, the file `siunitx.sty` is the file that should be analyzed.

Therefore, instead of focusing solely on `.dtx` files, it might be more practical to extend support to `.ins` files. We can implement `.ins` file support by having `explcheck` work similarly to DocStrip. The tool would transparently generate the contents of the files specified by an `.ins` file in-memory, process them individually, and then map the warnings and reports back to the original `.dtx` files. 

In the console output, the filename of the generated file (e.g., `siunitx.sty`) would appear at the top level, potentially alongside the `.ins` file (e.g., `siunitx.ins//siunitx.sty`) for better context. This makes the output more explainable (e.g., "I specified `siunitx.ins` as the input file—why am I seeing `siunitx.sty` in the output?"). Line-specific warnings and errors would be attributed to the relevant lines in the original `.dtx` files. File-wide warnings and errors would display the name of the output file (e.g., `siunitx.ins//siunitx.sty`) instead. For example:

```
$ explcheck siunitx.ins
```
```
Checking 1 file

Checking siunitx.ins//siunitx.sty    2 warnings

    siunitx-unit.dtx:123:1:    w123 some warning that points to a specific line
    siunitx.ins//siunitx.sty:  w456 some file-wide warning

Total: 0 errors, 2 warnings in 1 file
```

If we don't want to process all files generated by an `.ins` file, users could either specify the file(s) they want to process directly on the command line (e.g., `$ explcheck siunitx.ins//siunitx.sty`) or define which generated files to analyze or ignore using glob patterns in a configuration file.

**Comment by Witiko** [comment]
@josephwright Your feedback would be greatly appreciated. This feature is not part of the initial requirements, and I’m uncertain if I’ll be able to include it in the initial stable release, planned for next September as per the project proposal. However, ironing out the requirements now increases the likelihood that, if not me, then a contributor could pick this up in the future.

**Comment by Witiko** [comment]
Dear Frank,

Currently, .dtx files are not directly supported because they may contain
fragments from multiple sources, making it difficult to statically analyze
and determine which parts belong together. Therefore, the tool should be
currently run on the .tex, .sty, and .cls files generated by DocStrip, not
the .dtx sources.

I am discussing ways to tackle the support for .dtx files (or perhaps more
promisingly, .ins files) with Joseph Wright on GitHub
<https://github.com/witiko/expltools/issues/8#issuecomment-2537390144>.
Until such support has been implemented, I plan to modify explcheck to
explicitly reject .dtx files with an informative message, so that it's
clear to users that .dtx files are not a suitable target for the analysis.

Regarding your point about comments, you're right; the preprocessing step
should not raise error E102 (expl3 control sequences in non-expl3 parts)
when encountering things that resemble expl3 control sequences within TeX
comments. However, in this specific case, it simply reflects that explcheck
is confused about which parts of the .dtx file are or are not part of expl3.

Regarding the output format: It is inspired by the output of luacheck and,
as far as I know, not machine-readable from emacs. Instead, the output is
intended for human eyes. For text editors, an LSP (language server
protocol) support is envisioned in a future release. Furthermore, the
command-line interface, implemented in the file explcheck-cli.lua, is just
a thin wrapper over the explcheck-*.lua modules. Therefore, producing a
different output format is not a significant challenge.

Best,
Vit

**Comment by Witiko** [comment]
Good morning Vit

> Currently, .dtx files are not directly supported because they may 
> contain fragments from multiple sources, making it difficult to 
> statically analyze and determine which parts belong together. Therefore, 
> the tool should be currently run on the .tex, .sty, and .cls files 
> generated by DocStrip, not the .dtx sources.

make sense, but it is in fact interesting to pick up names in 
documentation, because often enough they contain human errors, such as 
misspellings of functions and those are hard to spot for a human but 
perhaps much easier for a machine.
> 
> Regarding your point about comments, you're right; the preprocessing 
> step should not raise error E102 (expl3 control sequences in non-expl3 
> parts) when encountering things that resemble expl3 control sequences 
> within TeX comments. However, in this specific case, it simply reflects 
> that explcheck is confused about which parts of the .dtx file are or are 
> not part of expl3.

being a comment, I guess it could contain both 2e or expl3 code 
regardless of the surrounding, e.g., in 2e code you might have a comment 
such as "% similar to \file_forget:n" inside 2e and it shouldn't raise 
an error.

Same really with the .dtx comments where the documentation is likely to 
talk about 2e and expl3 names

> Regarding the output format: It is inspired by the output of luacheck 
> and, as far as I know, not machine-readable from emacs. Instead, the 
> output is intended for human eyes. For text editors, an LSP (language 
> server protocol) support is envisioned in a future release. Furthermore, 
> the command-line interface, implemented in the file explcheck-cli.lua, 
> is just a thin wrapper over the explcheck-*.lua modules. Therefore, 
> producing a different output format is not a significant challenge.

ok, I'm pretty sure there is something to directly support that format. 
I think Nelson Beebe once mentioned that to me ... going to ask him

thanks
frank

**Comment by Witiko** [comment]
Good evening, Frank,

Thank you for sharing that—it’s quite interesting! As a user of Vim (the
superior editor), I wasn’t aware that this format is considered a de facto
standard (?) for what I assume is primarily compiler output.

That said, there are challenges beyond indentation. Specifically, explcheck
shortens long pathnames by iteratively replacing path segments with "/.../"
until the pathname is short enough, or until it runs out of segments. In
the latter case, explcheck replaces the prefix of the filename itself with
"...". While this makes the output more human-readable, it’s likely to
confuse Emacs.

To address this, I think the best approach would be to add a new
command-line option, -p or --porcelain, to generate machine-readable output
in a format compatible with Emacs.

Let me know your thoughts!

Best regards,
Vit

**Comment by FrankMittelbach** [comment]
Well, it is generated by unix tools such as grep for instance. Try `grep -n searchstring *`. With this command you have to explicitly ask for line numbers with `-n`but other tools do it automatically and it makes sense as a simple index into multiple file positions. Perhaps even the "superior editor" might have a way to make direct use of such grep output :-) the operating system without editor (aka emacs) can.

As to machine readable, yes, why not have some option (might even be `-n`(numbers) or `-m`(machine) and in that case just don't shorten things but leave path and name alone. And also drop the surrounding text, that you currently put before and after and output only the findings.

**Comment by Witiko** [comment]
Seems easy enough. I will look into it later today.

**Comment by FrankMittelbach** [comment]
Actually, emacs doesn't care about extra lines not starting with that syntax. Don't know about other tools/editors that would understand the file:line:column message lines.

**Comment by Witiko** [comment]
> Well, it is generated by unix tools such as grep for instance. Try `grep -n searchstring *`. With this command you have to explicitly ask for line numbers with `-n`but other tools do it automatically and it makes sense as a simple index into multiple file positions.

Line numbers are common across different tools but column numbers seem more unusual, which is why my mind went to compilers.

**Comment by Witiko** [comment]
> ok, emacs understands the syntax file:line:column message but only if the "file" starts at the beginning of the line then you can do <meta> M compile explcheck filename and it opens a buffer with the output and you can point and click to get to each position ... really handy, but if fails because it reads the spaces at the beginning of the line as part of the file name to open

> Actually, emacs doesn't care about extra lines not starting with that syntax. Don't know about other tools/editors that would understand the file:line:column message lines.

I believe you should be able to parse the output in Vim in a similar way with the following configuration:

```
set makeprg=explcheck\ %
set errorformat=%f:%l:%c\ %m
```

Then, you would:
1. Open a file in Vim.
2. Run `:make`. Vim will execute the command specified in `makeprg` (e.g., `explcheck <current-file>`).
3. Vim will populate the Quickfix list with the output from explcheck, based on the `errorformat`.

Navigating results:
- To open the Quickfix list: `:copen`
- To jump to the next error: `:cnext`
- To jump to the previous error: `:cprev`

**Comment by Witiko** [comment]
> Until such support has been implemented, I plan to modify explcheck to explicitly reject .dtx files with an informative message, so that it's clear to users that .dtx files are not a suitable target for the analysis.

I forbade checking .ins and .dtx files in #14.

> Seems easy enough. I will look into it later today.

I implemented the command-line option `--porcelain` in #15. I have yet to test the output with Vim and emacs.

Both changes will be released in explcheck v0.2.0, potentially as soon as later today.

**Comment by Witiko** [comment]
> Both changes will be released in explcheck v0.2.0, potentially as soon as later today.

I just realized that there is another issue in this ticket that should also be addressed:

> > On the whole I would assume one can take % as a comment char for granted (and perhaps also "^^A" in a .dtx file) but in any case, I think it would be nice if one could specify whether or not comments should be be parsed.
>
> Regarding your point about comments, you're right; the preprocessing step should not raise error E102 (expl3 control sequences in non-expl3 parts) when encountering things that resemble expl3 control sequences within TeX comments. However, in this specific case, it simply reflects that explcheck is confused about which parts of the .dtx file are or are not part of expl3.

Therefore, I opened a draft pull request https://github.com/Witiko/expltools/pull/16 that documents the issue using two failing unit tests. I will look into implementing a more robust parser of TeX comments and making the tests pass later. This does not block today's release of explcheck v0.2.0, though.

> > Regarding your point about comments, you're right; the preprocessing step should not raise error E102 (expl3 control sequences in non-expl3 parts) when encountering things that resemble expl3 control sequences within TeX comments.
>
> make sense, but it is in fact interesting to pick up names in documentation, because often enough they contain human errors, such as misspellings of functions and those are hard to spot for a human but perhaps much easier for a machine.

I'm not currently 100% clear on what exactly the validation of comments in .dtx files would entail; we'd need to analyze the requirements. After I have closed this ticket in pull requests #15 and #16, I will open a separate ticket specifically for the support of .ins and .dtx files. We can continue with the ideas there.

> > To address this, I think the best approach would be to add a new command-line option, -p or --porcelain, to generate machine-readable output in a format compatible with Emacs.
>
> As to machine readable, yes, why not have some option (might even be -n (numbers) or -m (machine) and in that case just don't shorten things but leave path and name alone.

I am happy to add aliases other than --porcelain as well as the option to specify the desired output in a config file in due time. However, at this early stage of development, I'd rather experiment and see which ideas work and which don't. When it's clear that the output produced by the option --porcelain will stay, then we can nail the API down.

**Comment by FrankMittelbach** [comment]
> > 
> 
> We can add aliases other than --porcelain as well as the option to specify the desired output in a config file eventually. However, at this early stage of development, I'd rather experiment and see which ideas work and which don't. When it's clear that the output produced by the option --porcelain will stay, then we can nail the API down.

I wasn't  suggesting to have several aliases, all I was suggesting that those couldd be alternatives. Anyway, an additional short option like `-p`as you originally suggested, would be nice in my opinion.
